colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -0,0 +1,838 @@
+            optimizer (OptimizerWrapper): Optimizer to update the model
+            input_obj (Optional[dict]): x.
+            output_obj (Union[dict, torch.Tensor]): y.
+            output_obj_grad (dict): dy.
+
+        Returns:
+            Optional[dict]: dx.
+        """
+        # calculate bwd b step ; only dx = w*dy;
+
+        # Retain the grad on the input_obj.
+        tree_map(retain_grad, input_obj)
+
+        if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+            # loss backward; output_obj is loss
+            output_obj_grad = None
+        optimizer.backward_by_grad(
+            tensor=output_obj,
+            grad=output_obj_grad,
+            inputs=input_obj,
+            retain_graph=True,
+        )
+        return input_obj.grad
+
+    def backward_w_step(
+        self,
+        model_chunk: Union[ModuleList, Module],
+        model_chunk_id: int,
+        optimizer: OptimizerWrapper,
+        output_obj: Union[dict, torch.Tensor],
+        output_obj_grad: Optional[dict],
+    ):
+        """Backward dw step of the pipeline; we calculate "dw = x*dy" here;
+
+        Args:
+            model_chunk (ModuleList or Module): Model Chunk to be run;
+            model_chunk_id (int): The current model chunk idx;
+            optimizer (OptimizerWrapper): Optimizer to update the model
+            output_obj (Union[dict, torch.Tensor]): y.
+            output_obj_grad (dict): dy.
+
+        Returns:
+            Nothing need to return; we only calculate dw then update w;
+        """
+        # calculate bwd w step ; only dw = x*dy;
+
+        if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
+            # loss backward; output_obj is loss
+            output_obj_grad = None
+        optimizer.backward_by_grad(
+            tensor=output_obj,
+            grad=output_obj_grad,
+            inputs=list(model_chunk[model_chunk_id].parameters()),
+            retain_graph=False,
+        )
+        # if model_chunk_id == 0:
+        #     optimizer.backward_by_grad(
+        #         tensor=output_obj,
+        #         grad=output_obj_grad,
+        #         inputs=list(model_chunk[model_chunk_id].parameters()),
+        #         retain_graph=False,
+        #     )
+
+        # else:
+        #     if self.stage_manager.is_first_stage(ignore_chunk=True):
+        #         optimizer.backward_by_grad(
+        #             tensor=output_obj,
+        #             grad=None,
+        #             inputs=list(model_chunk[model_chunk_id].parameters()),
+        #             retain_graph=False,
+        #         )
+        #     else:
+        #         optimizer.backward_by_grad(
+        #             tensor=output_obj,
+        #             grad=output_obj_grad,
+        #             inputs=list(model_chunk[model_chunk_id].parameters()),
+        #             retain_graph=False,
+        #         )
```

Comment:
remove useless comments

Reply:
> remove useless comments

Removed in https://github.com/hpcaitech/ColossalAI/pull/6034/commits/e6e1a97a6d2d69fc8cd2907883e0627a61e6f372;